src/openrct2-ui/windows/Research.cpp

```diff
@@ -154,117 +113,218 @@ static constexpr const StringId ResearchStageNames[] = {
-}
+        Invalidate();
+        if (p == WINDOW_RESEARCH_PAGE_DEVELOPMENT)
+        {
+            width = WW_DEVELOPMENT;
+            height = WH_DEVELOPMENT;
+        }
+        else
+        {
+            width = WW_FUNDING;
+            height = WH_FUNDING;
+        }
+        WindowEventResizeCall(this);
+        WindowEventInvalidateCall(this);
 
-#pragma region Development page
+        WindowInitScrollWidgets(*this);
+        Invalidate();
+    }
 
-/**
- *
- *  rct2: 0x006B6B38
- */
-static void WindowResearchDevelopmentMouseup(WindowBase* w, WidgetIndex widgetIndex)
-{
-    switch (widgetIndex)
+    void OnUpdate() override
     {
-        case WIDX_CLOSE:
-            WindowClose(*w);
-            break;
-        case WIDX_TAB_1:
-        case WIDX_TAB_2:
-            WindowResearchSetPage(w, widgetIndex - WIDX_TAB_1);
-            break;
-        case WIDX_LAST_DEVELOPMENT_BUTTON:
-            News::OpenSubject(News::ItemType::Research, gResearchLastItem->rawValue);
-            break;
+        // Tab animation
+        if (++frame_no >= window_research_tab_animation_loops[page])
+            frame_no = 0;
+        InvalidateWidget(WIDX_TAB_1 + page);
     }
-}
 
-/**
- *
- *  rct2: 0x0069CBA6
- */
-static void WindowResearchDevelopmentUpdate(WindowBase* w)
+    void OnMouseDown(WidgetIndex widgetIndex) override
+    {
+        if (page == WINDOW_RESEARCH_PAGE_FUNDING)
+        {
+            WindowResearchFundingMouseDown(this, widgetIndex, WIDX_RESEARCH_FUNDING);
+        }
+    }
+
+    void OnMouseUp(WidgetIndex widgetIndex) override
+    {
+        // Switch tab or close
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+            {
+                Close();
+                break;
+            }
+            case WIDX_TAB_1:
+            case WIDX_TAB_2:
+            {
+                SetPage(widgetIndex - WIDX_TAB_1);
+                break;
+            }
+        }
+
+        // Process mouse up for specific tab
+        switch (page)
+        {
+            case WINDOW_RESEARCH_PAGE_DEVELOPMENT:
+            {
+                WindowResearchDevelopmentMouseUp(widgetIndex, WIDX_CURRENTLY_IN_DEVELOPMENT_GROUP);
+                break;
+            }
+            case WINDOW_RESEARCH_PAGE_FUNDING:
+            {
+                WindowResearchFundingMouseUp(widgetIndex, WIDX_RESEARCH_FUNDING);
+                break;
+            }
+        }
+    }
+
+    void OnDropdown(WidgetIndex widgetIndex, int32_t selectedIndex) override
+    {
+        if (page == WINDOW_RESEARCH_PAGE_FUNDING)
+        {
+            WindowResearchFundingDropdown(widgetIndex, selectedIndex, WIDX_RESEARCH_FUNDING);
+        }
+    }
+
+    void OnPrepareDraw() override
+    {
+        auto* targetWidgets = window_research_page_widgets[page];
+
+        if (widgets != targetWidgets)
+        {
+            widgets = targetWidgets;
+            WindowInitScrollWidgets(*this);
+        }
+
+        disabled_widgets = 0;
```

Comment:
Where did this come from?

Reply:
Can't find a reason for it, good catch. I will remove it.